v2/lints/cabf_br/lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included.go

```diff
@@ -0,0 +1,64 @@
+/*
+ * ZLint Copyright 2020 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v2/lint"
+	"github.com/zmap/zlint/v2/util"
+)
+
+type OCSPIDPKIXOCSPNocheckExtNotIncluded struct{}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) Initialize() error {
+	return nil
+}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) CheckApplies(c *x509.Certificate) bool {
+	if util.HasEKU(c, x509.ExtKeyUsageOcspSigning) {
+		return true
+	}
+	return false
+}
+
+func (l *OCSPIDPKIXOCSPNocheckExtNotIncluded) Execute(c *x509.Certificate) *lint.LintResult {
+	if !util.HasEKU(c, x509.ExtKeyUsageOcspSigning) {
+		// This point should never be reached, because this LINT only applies to certificates with this EKU
+		return &lint.LintResult{Status: lint.Fatal}
+	}
+
+	if !util.IsExtInCert(c, util.OscpNoCheckOID) {
+		if util.IsServerAuthCert(c) {
+			// If the certificate is a TLS certificate, it is clear, that the BRGs apply and we have an ERROR
+			return &lint.LintResult{Status: lint.Error}
+		}
+
+		// If the certificate is not a TLS certificate, the BRGs apply, if one of the sibling certificates or the parent certificate has the Server Auth EKU
+		return &lint.LintResult{Status: lint.Warn, Details: "Check the sibling and parent certificate for the Server Auth EKU. If one of them contains the Server Auth EKU, this is an ERROR and not a WARN"}
```

Comment:
I'm not sure I understand this part of the body.

BRs 4.9.9 apply to OCSP Responder Certificates, independent of any additional EKUs at play here.

My best guess here is that this is trying to cover a scenario where a Root issues Intermediate 1 and Intermediate 2. Intermediate 1 has `id-kp-serverAuth` and Intermediate 2 has `id-kp-emailProtection`.

You're wanting to make sure this lint runs for the OCSP Responder for Intermediate 1, and I'm guessing that this is trying to exclude Intermediate 2's OCSP Responder, right? With the assumption being that the issuing CA runs all lints (and all lint categories) for all certificates within their issuance hierarchy (e.g. they lint all of their S/MIME certificates just like they lint their TLS certificates, it's just we don't have S/MIME lints).

Just wanting to double check my understanding so far - did I get this right?

Reply:
@sleevi Yes, you are totally correct. Let me try to give an example. Basically I want to achieve the following results:

https://crt.sh/?id=174851777&opt=zlint --> should be flagged as "Error"
https://crt.sh/?id=3065731585&opt=zlint --> should be flagged as "Pass"
https://crt.sh/?id=6487721485&opt=zlint --> should be flagged as "Warn" because one would need to check the intermediate CA above (https://crt.sh/?id=21606057) to see if it is capable of issuing TLS certificates and therefore the BRGs apply or not

I have tried to cover those three examples in https://github.com/RufusJWB/zlint/blob/master/v2/lints/cabf_br/lint_ocsp_id_pkix_ocsp_nocheck_ext_not_included_test.go

I hope my understanding of this whole issue is correct.

Reply:
Gotcha. Yeah, this is a tricky case!

We don’t have issuer context presently, which was something we discussed when implementing the checks to make sure the signature hash algorithm of a subscriber cert is appropriate for the issuing CAs private key (e.g. SHA-384 with P-256). While we could fake a little there, based on the size of r/s, we don’t have that same luxury here.

Under today’s model, it’ll require a human to sort out. If we warned/errored on everything, that would reflect that in the BRs it’s a MUST, but we know it would have a false positive like in the case of Siemens. For a case like that, the only way to know whether it’s a false positive or not would be to inspect all certificate paths, or at least, all known/possible issuer certs as inputs to this cert. If we made it an error, CAs would get in the habit of ignoring errors, or needing to suppress errors, and we don’t want to encourage that.

I’m mixed on 43-46, and wondering if it may be unnecessary. The reason is that it seems like it’s (potentially) linting a different property, or would at least be covered by that property, of not having a server cert be an OCSP responder cert to begin with. However, I think I can also see the argument being made here, which is that such a cert would be unambiguously in scope, and so it doesn’t hurt to make that `Fatal` when we’re for sure.

I think my substantive feedback is more that we should document this more, as it at least wasn’t immediately clear and obvious on the first pass, but assuming I’ve got that understanding right, I think you’re right and that it makes sense as the best we can do here. I do think we can also perhaps structure this a little easier; for example, we know unambiguous if the extension **is** present, then it’s a `PASS`, so we could return that result then, which would let us unnest this logic. That would make it easier to reason about the state tree, and also I suspect make it easier to document.

As it’s late here, I’ll pick up tomorrow with concrete code suggestions, unless this all makes sense and you wanted to give a go at a restructuring and updated documentation.

If you do decide to update, one big thing that threw me was the remark about ‘sibling cert’, since it’s not the sibling that decides, but whether the parent is “technically capable”. I’ll try to come up with some more concrete suggestions tomorrow, if that doesn’t immediately make sense.

Reply:
Thank you for your feedback.

1. Regarding the `FATAL` I don't think it fits here because acc. to https://github.com/zmap/zlint/blob/master/CONTRIBUTING.md#adding-new-lints `FATAL` is reserved for 

> when there is an unresolvable error in `zlint`

Therefore I would propose to keep it as `Error`

2. Regarding the proposed restructuring, I did try to reflect it in my commit https://github.com/RufusJWB/zlint/commit/d49bcea722c2a909b7a4b17ebe2b5abdea487498 . Do you think the code is better to understand now?